### ISSUE TITLE
How to mitigate infinite redraw loop

### ISSUE THREAD

**Issue by barneycarroll** [lhorie/mithril.js#439]
I have an issue I touched upon in this thread about [synchronous promise resolution](https://github.com/lhorie/mithril.js/issues/436). My specific problem is that I've entered a situation where a redraw is triggered inside the redraw loop. I thought I could sidestep this issue by using start / end computation hooks before running the code that calls the offending `promise.then( redraw )`:

```javascript
m.startComputation();
setTimeout( m.endComputation );
```

This didn't do what I was expecting — I ended up using this hack instead:

```javascript
var realRedraw = m.redraw;

function fakeRedraw(){};

for( var key in realRedraw ) fakeRedraw[ key ] = realRedraw[ key ];

function redrawHiatus(){
	m.redraw = fakeRedraw;

	setTimeout( function rebindRedraw(){
		m.redraw = realRedraw;
	} );
}
```

Is this a bug or a misunderstanding? 

Is it worth putting in safeguards to prevent redraw recursion or is this fair game for user error?

**Comment by barneycarroll** [comment]
Taking note of your point about setTimeout delays, would it be possible to re-bind redraw internally and reinstate it after build execution? There's plenty of rebinding going on Mithril core. Or am I missing something?

**Comment by lhorie** [comment]
It's a bit more complicate than that. The problem, as it stands, is that having controllers that are instantiated in the view (via Mithril.Elements, m.tags, modulator, whatever) means that they run after the view started getting constructed. This means changing m.redraw.strategy could break rendering (and also creates the m.redraw infinite loop).

I'm thinking I'll need to write a controller gathering recursive step in the `build` function so that controllers are guaranteed to run before rendering, much like config is guaranteed to run after rendering.

**Comment by barneycarroll** [comment]
I get the setTimeout thing — it's a proof-of-concept shortcut I wouldn't be using if I weren't monkey-patching on principle — nobody deserves arbitrarily deferred rebinds in otherwise reliable core APIs.

Setting that aside then, mixed redraw strategies in a single pass sounds like we're straying into multi-tenancy territory, inasmuch as core functions which were hitherto singletons need distinct execution contexts — except that in this situation we're potentially catering for inherently hierarchical structures (whereas multiple `m.route` executions at least get the privation of contrived use case), and hence potentially recursive execution contexts. That could dent performance. Moving on…

A lower order scope may dictate full redraw strategy and have that directive ignored until the next pass because the strategy change wasn't made before the current redraw pass. If you ask me, this is a use case that should be handled exclusively by `build` and specified in the VDOM notation — which is already possible via subtree, id and key. I've raised a ticket about cases where that doesn't cut the mustard, but if people were enabled to at least distinguish between these things we'd be in happier straights:

Too many of the current attempts at intuitive componentization lazily conflate the logic behind the build diff algorithm with a suitable model for module instance identification. The ambiguity there risks growing and I submit that DOM redraw policy can already be determined by controllers passing flags down through to views which can in turn provide user flags for subtree, key, id etc.…and should therefore be untouchable outside of that context.

Way back when I'd just started using Mithril I raised a ticket because I was surprised at redraw strategy behaviour needlessly blowing up my DOM by default every time I changed route, but then I got screwed when I switched it to 'diff' and was confused by views inheriting the `isInitialized` and `context` from 'components' that were entirely unrelated.

Making it clear that DOM build and module ie controller instantiation are separate is a necessary burden AFAIC. With the modulator API, this is only ever a problem if the user gives a shit, but they already have all the controls should they wish them.

**Comment by philtoms** [comment]
I don't think running though the controllers pre-build will help much unless you include view results. The problem is the order in which views get added to the virtual dom. In m.tags (and I guess m.module) this happens in the build phase as you say. In Mithril.Elements, component injection (controller and view) is rolled into the top level view construction. No matter how deep that construction goes, it pops itself off the stack to the same point that a standard mithril app would be at when it enters the build phase. In this scheme, it's impossible for components to have recursive adverse effects so config / redraw etc behaves as expected.

The full consequence of this did not hit me until I converted m.elements to m.tags in the MSK. In the m.tags version, knowledge of view construction order became a component design issue.

**Comment by lhorie** [comment]
@barneycarroll it's not about mixing redraw strategies. The problem is that redraw strategy doesn't work at all if a component is mounted from a view.

@philtoms yeah, notice I said controllers need to be guaranteed to run before *rendering* (as in, to the DOM), but not necessarily before virtual dom tree construction. I've been under the impression that merely passing a function as an argument would be enough to control order of construction, but I haven't yet put together a proof of concept to validate my assumption. Which component was giving you a hard time during the conversion?

**Comment by lawrence-dol** [comment]
I've dealt with this possibility by creating a "request redraw" utility:

    function redrawRqd() {
        if(!MithrilUtil.redrawPending) {
            MithrilUtil.redrawPending=true;
            setTimeout(function() {
                m.redraw();
                MithrilUtil.redrawPending=false;
                },20);                                                                                  // 1000/20 = 50 fps
        }
    }

Which queues a single asynchronous request to redraw at the next UI event-loop cycle and coalesces any overlapping requests. 

Any asynchronous action which changes the model can request a redraw. I use this on things like `m.request` and don't use start/end computation because I think it's plain wrong to suppress UI rendering until some arbitrary action or actions complete. 

Start/end computation effectively block the UI thread, from the point of view of the application's user.

**Comment by barneycarroll** [comment]
> @barneycarroll it's not about mixing redraw strategies. The problem is that redraw strategy doesn't work at all if called from a component that is mounted from a view.

If your top level controller sets one redraw strategy (or sticks with the default) and a lower-order controller requests a different redraw strategy, that is a mixed message. What my post is saying is that being able to change redraw strategy in a controller barely makes sense in the first place. The fact you currently have to do it this way is an implementation accident.

Or are you telling me current componentization implementations are broken because they don't allow a select2 plugin to dictate that the whole page must render from scratch when it's first initialised?

> I've dealt with this possibility

Hey @lawrence-dol you should have a look at [this](https://gist.github.com/barneycarroll/0e3c1b9811b47c012b13#file-modulator-js-L16-L42).

**Comment by philtoms** [comment]
> ..controllers need to be guaranteed to run before *rendering*

@lhorie, yep I understood that. Using a function to invert the construction order won't work because the virtual dom is not complete when rendering begins. The component won't be in the virtual dom until its reference node is traversed. This was the takeaway from the conversion.

I had to modify the tabset and the occlusionScroller. The tabset code was actually nicer because it removed the inverting function. 

@barneycarroll maybe I'm missing something but if the logic that leads to your synchronous promise hasn't changed between cycles, wouldn't that cause the loop irrespective of render order or strategy?

**Comment by lhorie** [comment]
> If your top level controller sets one redraw strategy (or sticks with the default) and a lower-order controller requests a different redraw strategy, that is a mixed message

Right, and again, it's not about mixing strategies. The scenario that breaks is e.g. 

```javascript
//say you want your app looks like this
Layout([
  StaticMenuWithExpensivePlugin() //and you want to change strategy here because you think it makes more sense than changing in the root component
  DynamicContent()
])
```

If you were to keep setting strategies, the last call should win (although I can't think why anyone would want to keep changing m.redraw.strategy, since that is equivalent to setting a variable multiple times).

**Comment by barneycarroll** [comment]
@philtoms the controller instance persists across redraws.

> I can't think why anyone would want to keep changing m.redraw.strategy

@lhorie exactly, it's a straw man. If `StaticMenu` asks for `diff` but `DynamicContent` asks for `all`, it's unexpected and undesirable for `StaticMenu` (actually, the entire app) to render from scratch. Hence my select2 plugin analogy. Imagine `Layout` is itself a lower-order controller – the deeper you go, the crazier it becomes that a leaf node module can decide to regenerate the entire layout DOM from scratch because its author is lazy about init logic. 

What is desirable (and intuitive) is for a module to take responsibility of its own view and say "**I** should render from scratch". Here's the naive way to do that:

```javascript
let instanceKeys = 0;

let DynamicContent = {
  controller(){
    this.key = instanceKeys++;
  },
  view( ctrl ){
    return m( 'canvas', {
      key    : 'DynamicContent' + ctrl.key
    } );
  }
};
```

This is still sub-par. The more you think about it, the more you realise:

* This is all view logic – the controller logic and higher scope should be unnecessary
* I believe you've solved automatic context-determination in m.components – we could use that feature in along with '[redraw as a VDOM attribute](https://github.com/lhorie/mithril.js/issues/409)' to bake this logic into `build`.
* Do I even want a fresh DOM tree? Mithril's diff engine is awesome, it's why I'm using it in the first place – I don't want to hack it or sidestep it. More likely the thing I'm really after is a hook for a 3rd party plugin called in config to do proper setup and teardown at the right times.
* Globally unique primitive identifiers are wasteful & imprecise

```javascript
let DynamicContent = {
  view( ctrl ){
    return m( 'canvas', {
      config : function( el, init, context ){
        if( !init || context.ctrl !== ctrl ){
          if( context.ctrl ) while( el.lastChild ) el.removeChild( el.lastChild );

          setup();

          context.ctrl = ctrl;
        }
      }
    } );
  }
};
```

I *am* that guy with a large app containing modules within modules within modules, with RESTful route changes, continuous heavy data requests, dozens of expensive Highcharts instances with their own esoteric logic, and concurrent unrelated widgets X and Y that don't want to regenerate their hundred-odd DOM elements every time I can't be bothered to figure out re-initialization logic when I invoke plugin Z.

**Comment by lhorie** [comment]
> If StaticMenu asks for diff but DynamicContent asks for all, it's unexpected and undesirable for StaticMenu (actually, the entire app) to render from scratch

Again, the use case I'm talking about only changes strategy once. The point is that cut+pasting that line from one controller and pasting into another controller breaks. It's roughly equivalent in badness to a hypothetical situation where you cut and paste a m() call from a view to another, and the framework stops working for no particular reason.

FWIW, leaving it broken is easy. I just need to ignore it. But it doesn't change the fact that it's broken.

I've been thinking about #409 and it seems like what's missing is an "all-except-this" flag, where you can specify that something shouldn't be recreated from scratch even if strategy is "all".

**Comment by barneycarroll** [comment]
> Again, the use case I'm talking about only changes strategy once 

Because the 'use case' is just a validation of the current API and doesn't correspond to a credible user story. The example happens to have trivial complexity, and doesn't match up to real world concerns. IRL an app will have a multitude of components that, according to current Mithril API, may elect to use `m.redraw.strategy`, and these could be nested in any configuration.

Let's make this intended usage slightly more credible:
```
cApp
  cHeader
  cDataViewer
    cDataSources.map( source => cParameters
      cHighchart
  cChatRoom
```

`cParameters` contains a lot of DOM for a GUI that configures the range and parameters of a data source interface. These are all expressed in 'pure' Mithril with limited to no config. There's no need to destroy and recreate these elements every time a new source comes in – so `cParameters`' controller switches to `diff`. `cHighchart` espouses the current API's intended philosophy of `all` – it's primary role is to house a config'd element where setup / teardown just works easier by destroying the DOM and starting from scratch each time round. `cHighchart` will always execute after `cParameters` in invocation order, so actually `cParameters`'s `diff` request is always overridden. But this is all academic, because `cChatRoom` probably has its own opinions too. So redraw strategy is completely unreliable at any kind of scale.

At the risk of over-egging the point, *there is no use case* for globally changing `m.redraw.strategy` that isn't an inefficient and unreliable hack as a substitute for granular diff directives in the DOM itself – redraw strategy should always be `diff`. What users really want is the ability to assert controller instance freshness in the VDOM or config execution context (covered in the examples of my previous comment). The dubious case of `none` (can't think of any real application) is covered by `{ subtree }`.

**Comment by lhorie** [comment]
> So redraw strategy is completely unreliable at any kind of scale

That's a valid point. It's there for the case where the developer is the sole authority on whether redraw strategy is to be changed at all, but this isn't always realistic.

> The dubious case of none (can't think of any real application) is covered by { subtree }.

Not exactly. That's documented here: http://lhorie.github.io/mithril/mithril.redraw.html#preventing-redraws-on-events

**Comment by lhorie** [comment]
@philtoms sorry, I ignored you there for a bit. What problems do you run into when component view construction happens within `build` as opposed to at template construction time (as in Mithril.Elements)?

> it's impossible for components to have recursive adverse effects so config / redraw etc behaves as expected.

I can see unit testing views and template transformation functions would be problematic, and I get the part about redraw, but I don't understand the problem re: configs.

**Comment by philtoms** [comment]
> I don't understand the problem re: configs.

I was basing my observation on the fact that complete view construction before the build produces config behaviour consistent with mainstream mithril. So there would be no component related edge cases at all. A small point really.

**Comment by lawrence-dol** [comment]
@lhorie,

I expected Barney to call this out, but since he didn't...

> it's imperative for the developer to be the sole authority on whether redraw strategy is to be changed at all

I think that's entirely Barney's core premise -- said "developer" is entirely mythical in any application of non-trivial complexity. 

Your statement, quoted above, is the antithesis of my claim that "no one part of the application can possibly make, globally, positive claims about whether redrawing is necessary; they can only make it for their particular part of the UI in that their part of the UI reflects some part of the model which they know has changed (or at least probably has changed) since the last UI render.

In other words, there is no such "sole authority", even if the same developer wrote all the code. It's a little like saying the browser's page content can dictate to the browser that it's necessary to redraw the browser window frame, tabs, toolbar, status bar and other accoutrements -- the very idea warrants a "huh? why?" response.

**Comment by barneycarroll** [comment]
> no one part of the application can possibly make, globally, positive claims

@lawrence-dol I think a large proportion of the more complex issues with Mithril at present revolve around this. Internal heuristics that allow granular control from within nodes (components, controllers, views, elements) are a difficult problem for core development, but too often that difficulty is conflated with user expectation: redraw scope and instance management are difficult internally, so let's assume the user expects to manage these globally. If they want to do this with hierarchical scope, they're the exception to the rule and it's there responsibility to devise mechanisms that deal with this. It's the other way around: wanting to change refresh policy within any given scope is a credible concern – making global decisions is unituitive and dangerous, and only gets used at present because of core logic limitations.

**Comment by barneycarroll** [comment]
Having said that, I feel I must split hairs about the 'noredraw' thing. Stating that a particular branch of the DOM tree should always redraw, never redraw, redraw conditional to arbitrary complexity… that's legitimate. 

Limiting the scope of a redraw calculation is wrong. If you're worried that parts of the DOM may update that you don't want to, then your model or is screwed and your real problem is elsewhere. The only real reason for this is performance concern, which is not a qualitative thing. Mithril is by miles and away the most performant VDOM differ – complicating its core API to give more hooks to developers whose purpose doesn't solve any user problems except "I could save on memory by doing this" is the last thing Mithril should be looking to do. I must confess to a certain degree of contempt for developers who have yet to actually develop anything of any complexity and start suggesting optimisation hooks. If you're suffering from real performance problems, a diagnosis based on your application structure is in order.

**Comment by lawrence-dol** [comment]
@barneycarroll,

The TL;DR is that if modular components are to work, there can be no global flags controlling Mithril's behavior with respect to parts of the DOM outside of that for which a given component is responsible.

I raised this to draw attention to the problem that redraw strategy, as a global flag, is an extant problem for any application of non-trivial complexity. This is especially true for a modular app where the modules should have discrete concerns. Just like having a global name registry for modules. As it is now, redraw strategy is close to useless; I believe modules will make it completely so.

I was merely wanting to preempt an obvious issue and offer a proper conception based on my work solving a nearly identical problem for traditional GUIs where repainting parts of the display unnecessarily is also a problem for much the same reasons. (Though, Mithril's problem may actually harder because the things with which events are associated may be destroyed and recreated.)

> Limiting the scope of a VDOM redraw calculation is wrong – it's a similar hack to m.redraw.strategy

I agree; I have no practical use for redraw strategy except as a hack to suppress redrawing for input events who's effect has already been reflected in the GUI (per my previously reported problem of redrawing an input field or similar due to an input event or similar when the value was already changed before the event was triggered; which, BTW, caused input loss in IE for a while). And I do consider this to be a hack, but one for which there is currently no alternative.

> If you're suffering from real performance problems, a diagnosis based on your application structure is in order.

I was able to demonstrate real performance problems with a trivial fiddle due purely to excessive redrawing. Though, differently from use, I consider pegging the CPU unnecessarily a real problem even if my desktop system can (almost) keep up with it *because my system is not representative of my customer's systems*.
